---
title: "Examining "
author: "Sasha Mikheyev"
date: "4/28/2021"
output: pdf_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(scales)
library(ellipse)
library(gridExtra)
library(grid)
library(WGCNA)
# Also need the MASS library, which is not loaded because it conflicts with select
```

## *Protobothrops mucrosquamatus* (Taiwan habu)

```{r}
dir = list.dirs("agneesh/all_data/kallisto_results/", full.names = T, recursive = F)
filename = file.path(dir, "abundance.tsv")
names <- data.frame(name = filename) %>% separate(name, sep = "//", into = c(NA,"test")) %>% separate(test, sep = "/", into = c("name", NA)) %>% pull(name)

pm <- tibble(target_id = character(), est_counts = numeric(), tpm = numeric(), library = character())

for (j in seq(filename)) pm <- rbind(pm, read_tsv(filename[j]) %>% select(target_id, est_counts, tpm) %>% mutate(library = names[j]))
```

```{r taiwan, message=F}
genes <- read_tsv("ref/genes.txt", col_names = c("target_id", "gene"))
taiwan_venom <- read_rds("data/pm.RDS") %>% filter(grepl("Pm", library)) %>% left_join(genes, by = "target_id") %>% na.omit() %>% 
  group_by(gene) %>% summarize(tpm = sum(tpm))
taiwan_proteome <- read_csv("data/taiwan-secreted.csv", col_types = "-nc", col_names = c("gene", "type"))
rbind(read_tsv("data/1-Taiwan-Female/abundance.tsv") %>% mutate(library = "taiwan1"),
read_tsv("data/11-Taiwan-Female/abundance.tsv") %>% mutate(library = "taiwan11"),
read_tsv("data/2-Taiwan-Female/abundance.tsv") %>% mutate(library = "taiwan2"),
read_tsv("data/3-Taiwan-Male/abundance.tsv") %>% mutate(library = "taiwan3")) %>%
  left_join(genes, by = "target_id") %>% na.omit() %>% 
  group_by(library, gene) %>% summarize(tpm = sum(tpm), est_counts = sum(est_counts)) -> taiwan

averaged_data <- left_join(taiwan %>% group_by(gene) %>% summarize(tpm = sum(tpm), est_counts = sum(est_counts)), taiwan_venom %>% group_by(gene) %>% summarize(tpm = sum(tpm)), by = "gene") 
with(averaged_data, cor.test(tpm.x, tpm.y, method = "s"))
tp1 <- averaged_data %>% ggplot(aes(tpm.y, tpm.x, color = log10(est_counts ))) + geom_point() + scale_y_log10(labels=trans_format('log10',math_format(10^.x))) + scale_x_log10(labels=trans_format('log10',math_format(10^.x))) + theme_bw() + xlab("Venom gland RNA (TPM)") + ylab("Venom RNA (TPM)") + labs(color=expression('Log'[10]*" counts")) + theme(legend.position = 'bottom')

tp1

taiwan_merged_venom <- left_join(filter(taiwan, gene %in% taiwan_proteome$gene), taiwan_venom, by = "gene")

taiwan_merged_venom  %>% group_by(library)  %>% summarize(cor = cor(tpm.x, tpm.y), method = "s")
left_join(filter(taiwan, est_counts >= 1 & ! gene %in% taiwan_proteome$gene), taiwan_venom, by = "gene") %>% group_by(library)  %>% summarize(cor = cor(tpm.x, tpm.y), method = "s")

p1 <- taiwan_merged_venom %>% na.omit() %>% filter(est_counts >= 1) %>% ggplot(aes(tpm.y, tpm.x, color = library)) + geom_line(aes(group = gene), color = "grey") + geom_point() + scale_y_log10(labels=trans_format('log10',math_format(10^.x))) +scale_x_log10(labels=trans_format('log10',math_format(10^.x))) + theme_bw() + xlab("Venom gland gene expression (TPM)") + ylab("Venom mRNA level (TPM)") + guides(color = F) + ggtitle("Protobothrops mucrosquamatus") + theme(plot.title = element_text(face = "italic"))
p1
```

## Principal component analysis
```{r taiwan_pca}
taiwan_venom2 <- taiwan_venom <- read_rds("data/pm.RDS") %>% left_join(genes, by = "target_id") %>% na.omit() %>% 
  group_by(gene, library) %>% summarize(tpm = sum(tpm))

taiwan_merged2 <- rbind(taiwan %>% filter(est_counts >= 1) %>% select(-est_counts), taiwan_venom2) %>% pivot_wider(names_from = library, values_from = tpm) 

d <- taiwan_merged2 %>% ungroup %>% select(-gene,  -P_m_05_heart) 
d <- d[sort(rowSums(d, na.rm = T), decreasing = T, index.return = T)[[2]][1:5000],]
d <- log(d+1) %>% t() %>% dist()
fit <- MASS::isoMDS(d, k=2)
plotDat <- data.frame(x = fit$points[,1], y = fit$points[,2], tissue = c(rep("venom",4), rep(c("heart", "kidney", "liver"),2), c("kidney", "liver"), c("heart", "kidney", "liver"), rep("venom gland", 28)) ) 

centroids <- aggregate(cbind(x,y) ~ tissue, plotDat, mean)
conf.rgn  <- do.call(rbind, lapply(unique(plotDat$tissue), function(t)
  data.frame(tissue = as.character(t),
             ellipse(cov(plotDat[plotDat$tissue == t, 1:2]),
                     centre=as.matrix(centroids[centroids$tissue == t,2:3]),
                     level=0.95),
             stringsAsFactors=FALSE)))

tp2 <- ggplot(plotDat, aes(x,y,color = tissue)) + geom_point(size=3) + theme_bw() + geom_path(data=conf.rgn) + xlab("NMDS Axis 1") + ylab("NMDS Axis 2")+ theme(legend.position = 'bottom')
tp2

g <- grid.arrange(tp1, tp2, ncol = 1)
ggsave(file="plots/Figure 2 taiwan.pdf", g, width = 5, height = 10) 
```

As expected given the lower coverage the venom RNA nests with the venom gland samples

### Module preservation in Taiwan habu data
```{r modules, eval = F}
modules <- read_csv(file = "data/modules.csv")

taiwan_venom_matrix <-  rbind(read_rds("data/pm.RDS") %>% filter(grepl("Pm", library)) %>% select(-est_counts),
      rbind(read_tsv("data/1-Taiwan-Female/abundance.tsv") %>% mutate(library = "taiwan1"),
            read_tsv("data/11-Taiwan-Female/abundance.tsv") %>% mutate(library = "taiwan11"),
            read_tsv("data/3-Taiwan-Male/abundance.tsv") %>% mutate(library = "taiwan3")) %>% 
        filter(est_counts > 1) %>% select(target_id, library, tpm) %>%
        mutate(tpm = log(1+tpm)) %>% select(library, target_id, tpm)) %>% 
        pivot_wider(names_from = target_id, values_from = tpm)

taiwan_venom_matrix2 <- taiwan_venom_matrix[,intersect(modules$target_id, colnames(taiwan_venom_matrix)[colSums(is.na(taiwan_venom_matrix)) == 0])]
gsg <- goodSamplesGenes(taiwan_venom_matrix2)
gsg$allOK

goodModules <- modules %>% filter(target_id %in% colnames(taiwan_venom_matrix2)) %>% group_by(colors) %>% summarize(count = n() > 50) %>% filter(count == T) %>% pull(colors) # filter out small modules

keep <- intersect(modules %>% filter(colors %in% goodModules) %>% pull(target_id), colnames(taiwan_venom_matrix2)) 

setLabels <- c("old", "new");
multiExpr <- list(old = list(data = taiwan_venom_matrix2[1:28, keep]), new = list(data = taiwan_venom_matrix2[29:31, keep]))
multiColor <- list(old = modules %>% filter(target_id %in% keep) %>% pull(colors))

allowWGCNAThreads(10)
mp <- modulePreservation(multiExpr, multiColor,
                         parallelCalculation = T,
                         referenceNetworks = 1,
                         checkData = F,
                         nPermutations = 200,
                         randomSeed = 1,
                         verbose = 3)

mp$preservation$Z$ref.old$inColumnsAlsoPresentIn.new

# empirical p-value of turq
10^mp$preservation$log.p$ref.old$inColumnsAlsoPresentIn.new$log.psummary.pres[5]

```

The metavenom module is highly preserved in the venom RNA

## *Protobothrops flavoviridis* (Okinawa habu)

```{r okinawa, message=F}
okinawa_old <- read_tsv("old/okinawa.txt") # protein data from Aird et al.
rbind(read_tsv("data/7-Okinawa-Male/abundance.tsv") %>% mutate(id = "okinawa7"),
read_tsv("data/8-Okinawa-Female/abundance.tsv") %>% mutate(id = "okinawa8"),
read_tsv("data/9-Okinawa-Male/abundance.tsv") %>% mutate(id = "okinawa9") ) %>% 
  left_join(read_tsv("data/okinawa_ref/abundance.tsv")  %>% mutate(ref = tpm) %>% select(ref, target_id), by = c("target_id")) -> okinawa

okinawa_merged <- left_join(okinawa_old, okinawa, by =c("Pf" = "target_id"))

okinawa_merged %>% group_by(id) %>% filter(est_counts >= 0) %>% summarize(cor = cor(tpm, ref), method = "s")
cor.test(okinawa_old$fpkm, okinawa_old$`frag/len`)
okinawa_merged %>% group_by(id) %>% filter(est_counts >= 0) %>% summarize(cor = cor(tpm, `frag/len`), method = "s")

p2 <- okinawa_merged %>% filter(est_counts >= 1) %>% ggplot(aes(ref, tpm, color = id)) + geom_line(aes(group = Pf), color = "grey")+ geom_point() + scale_y_log10(labels=trans_format('log10',math_format(10^.x))) +scale_x_log10(labels=trans_format('log10',math_format(10^.x))) + theme_bw() + xlab("Venom gland gene expression (TPM)") + ylab("Venom mRNA level (TPM)") + guides(color = F) + ggtitle("Protobothrops flavoviridis") + theme(plot.title = element_text(face = "italic"))
p2
```

## *Ovophis okinavensis* (Hime habu)

```{r ovophis, message=F}
ovophis_old <- read_tsv("old/ovophis.txt") # protein data from Aird et al.
rbind(read_tsv("data/10-Hime-Female/abundance.tsv") %>% mutate(id = "hime10"),
      read_tsv("data/4-Hime-Female/abundance.tsv") %>% mutate(id = "hime4"),
      read_tsv("data/5-Hime-Female/abundance.tsv") %>% mutate(id = "hime5"),
read_tsv("data/6-Hime-Female/abundance.tsv") %>% mutate(id = "hime6") ) %>% 
  left_join(read_tsv("data/ovophis_ref/abundance.tsv")  %>% mutate(ref = tpm) %>% select(ref, target_id), by = c("target_id")) -> ovophis

ovophis_merged <- left_join(ovophis_old, ovophis, by =c("Oo" = "target_id"))

ovophis_merged %>% group_by(id) %>% filter(est_counts > 1) %>% summarize(cor = cor(tpm, ref), method = "s")
cor.test(ovophis_old$fpkm, ovophis_old$`frag/len`)
ovophis_merged %>% group_by(id) %>% filter(est_counts > 1) %>% summarize(cor = cor(tpm, `frag/len`), method = "s")

p3 <- ovophis_merged %>% filter(est_counts > 1) %>% ggplot(aes(ref, tpm, color = id))  + geom_line(aes(group = Oo), color = "grey") + geom_point() + scale_y_log10(labels=trans_format('log10',math_format(10^.x))) +scale_x_log10(labels=trans_format('log10',math_format(10^.x))) + theme_bw() + xlab("Venom gland gene expression (TPM)") + ylab("Venom mRNA level (TPM)") + guides(color = F) + ggtitle("Ovophis okinavensis") + theme(plot.title = element_text(face = "italic")) 

p3
```

```{r final_plot}
p <- list(p1,p2,p3) %>% map(~.x + labs(x=NULL, y=NULL))
grid.arrange(grobs=p, ncol=3, bottom = "Venom gland gene expression (TPM)", left = "Venom mRNA level (TPM)")

g <- arrangeGrob(grobs=p, ncol=3, bottom = "Venom gland gene expression (TPM)", left = "Venom mRNA level (TPM)") 
ggsave(file="plots/Figure 1 TPM.pdf", g, width = 10, height = 5) 
```